packages/engine-server/src/DendronEngineV3.ts

```diff
@@ -290,12 +296,328 @@ export class DendronEngineV3 implements DEngine {
     return resp.data ? resp.data : [];
   }
 
+  /**
+   * See {@link DEngine.writeNote}
+   */
+  async writeNote(
+    note: NoteProps,
+    opts?: EngineWriteOptsV2
+  ): Promise<WriteNoteResp> {
+    let changes: NoteChangeEntry[] = [];
+    let error: DendronError | null = null;
+    const ctx = "writeNewNote";
+    this.logger.info({
+      ctx,
+      msg: `enter with ${opts}`,
+      note: NoteUtils.toLogObj(note),
+    });
+
+    // Apply hooks
+    if (opts?.runHooks === false) {
+      this.logger.info({
+        ctx,
+        msg: "hooks disabled for write",
+      });
+    } else {
+      const hooks = _.filter(this.hooks.onCreate, (hook) =>
+        NoteUtils.match({ notePath: note.fname, pattern: hook.pattern })
+      );
+      const resp = await _.reduce<DHookEntry, Promise<RequireHookResp>>(
+        hooks,
+        async (notePromise, hook) => {
+          const { note } = await notePromise;
+          const script = HookUtils.getHookScriptPath({
+            wsRoot: this.wsRoot,
+            basename: hook.id + ".js",
+          });
+          return HookUtils.requireHook({
+            note,
+            fpath: script,
+            wsRoot: this.wsRoot,
+          });
+        },
+        Promise.resolve({ note })
+      ).catch(
+        (err) =>
+          new DendronError({
+            severity: ERROR_SEVERITY.MINOR,
+            message: "error with hook",
+            payload: stringifyError(err),
+          })
+      );
+      if (resp instanceof DendronError) {
+        error = resp;
+        this.logger.error({ ctx, error: stringifyError(error) });
+      } else {
+        const valResp = NoteUtils.validate(resp.note);
+        if (valResp instanceof DendronError) {
+          error = valResp;
+          this.logger.error({ ctx, error: stringifyError(error) });
+        } else {
+          note = resp.note;
+          this.logger.info({ ctx, msg: "fin:RunHooks", payload: resp.payload });
+        }
+      }
+    }
+
+    // Check if another note with same fname and vault exists
+    const resp = await this._noteStore.find({
+      fname: note.fname,
+      vault: note.vault,
+    });
+    const existingNote = resp.data ? resp.data[0] : undefined;
+    // If a note exists with a different id but same fname/vault, then we throw an error unless its a stub or override is set
+    if (existingNote && existingNote.id !== note.id) {
+      // If note is a stub or client wants to override existing note, we need to update parent/children relationships since ids are different
```

Comment:
are we making it a goal in the new engine to keep link state up to date?

Reply:
link state needs to be up to date for refactoring operations as well as deletion to work properly. so yes

Reply:
In which case, looks like we're missing some logic here?

Reply:
If we're referring to backlinks, I'm including that in the next pr as part of the `rename` refactoring since it includes some link modification logic